### ISSUE TITLE
Running GATK-HaplotypeCaller from Java

### ISSUE THREAD

**Issue by virtuabhi** [broadinstitute/gatk-protected#839]
Hi Everyone,

I have a use case where I need to directly call HaplotypeCaller from Java code. I can pass the required command line arguments (as String array). I am wondering how to call the main class with the required arguments.

I have tried calling org.broadinstitute.hellbender.Main with arguments but I get "HaplotypeCaller is not a valid command" (HaplotypeCaller works from the command-line prompt). Is this the correct class? Or should I create object of some other class and call its methods directly? Is there some way to load the list of available commands before hand? 

Thank you very much for your time.

Abhishek

**Comment by vdauwera** [comment]
Hi @virtuabhi, this is out of scope of what we normally support so I can't guarantee you'll get an answer. However, if you post a bit more detail (like, what the surrounding code looks like and what you're ultimately trying to do) our engineers may be able to suggest a solution.

**Comment by virtuabhi** [comment]
Hi @vdauwera, @lbergelson 

I spent some more time on the issue and I believe that the problem is at my end with some incompatible versions of libraries and environment variables. I will try to fix it and let you know if I get errors from the GATK side.

Thank you for your time.

**Comment by lbergelson** [comment]
@virtuabhi What are you trying to do exactly?  If you can explain your goal in a bit more detail I can probably help you accomplish it.  If you want to directly run an instance of HaplotypeCaller with an arg array then the easiest thing is probably:

```
  new HaplotypeCaller().instanceMain(args);
```

That's going to work better than trying to invoke `Main.main()` since that will eventually call `System.exit()` and kill your program.

**Comment by virtuabhi** [comment]
Thanks @lbergelson . I have added a new ReadsDataSource which reads files from HDFS in the GATK source code.

I edited the build file of GATK and added a dependency to Hadoop JAR files of version 2.5.2 (the version installed on my cluster). The GATK JAR already includes a dependency to Hadoop version 2.7.2. So I was ending up with a GATK JAR with libraries from two versions of Hadoop which was causing the error.

While finding the cause for this error, one of the things I tried was calling the main function from external GATK JAR using Java reflection and that's when I got the "command not found" error. However, calling the GATK-protected JAR from command line (as a different process) has always worked. Also, by importing the GATK JAR and calling the main function with an array of arguments, I was able to start the program.

I have fixed this issue now (changed all Hadoop versions to 2.5.2) and am calling GATK from command line which works without any issue. Will keep the information in mind about calling instanceMain to avoid the System.exit call.

Thank you for your help!
Abhishek
